### DISCUSSION TITLE
Is there an issue to follow to know when input-leap is ready for beta testing?

### DISCUSSION THREAD

**Discussion by axelson** [input-leap/input-leap#1793] [Q&A]
Right now on the [readme](https://github.com/input-leap/input-leap/blob/3e681454b737633a70f2f3b789046a5cb1946708/README.md#L20-L21) it states:

> At this time, Input Leap is in heavy development, and not ready for production use

Which is 100% understandable. What I'd like to know is how can I be notified when input-leap is either ready for beta testing (or production use).

My 2 cents is that it would be helpful to:
- Create a tracking issue (or possibly two, one for beta release and one for production release)
  - Maybe there is already one but I'm just having trouble finding it
- Pin that issue
- Mention that issue in the README

By the way I've been a Synergy/Barrier for a long time (and even tried my hand at development for a bit :face_in_clouds:) and I'm really excited to see development continue! :tada:

**Comment by ftoledo** [comment]
binary release for testing will be great!

**Comment by sithlord48** [comment]
# Binary builds are currently available
# Do NOT use them until you read this ENTIRE  post
   We have a CI system that does builds and uploads them we currently do not have a "release" step. You can still access the built binaries for every CI Action that was run. Each Action triggers two job for every `Pull Request` and every time we `land a Pull Request`. In order to be able to download the artifacts YOU MUST BE LOGGED IN TO GITHUB.
 
## Look at our [Actions](https://github.com/input-leap/input-leap/actions) tab in the repo to view all ci builds 

Each job is named after action that caused it. This is either a Pull Request's title or a Merge of a Pull Request #### .."
Looking on the right side you can see the branch that triggered the job in blue. Look for the name of the branch you want `master` is the "current" code. While you can try any proposed Pull Request build, I do not recommend this unless your testing that specific Pull Request.

Each of our action events triggers two different jobs with different artifacts produced. The jobs name is visible Just below the Actions Name
 - `Build Tests` is our general builder it makes the builds for windows, mac os and linux 
 - `Distribution package test builds` build RPM files only (maybe will in the future make others ? )

Example Action Pair
![Screenshot_20240224_091209](https://github.com/input-leap/input-leap/assets/7450820/9e934536-7f6a-43f1-ad7c-1ff092be556c)

Once you pick the build you want to view you will be a on  page that looks like [this one](https://github.com/input-leap/input-leap/actions/runs/8020393415)  Towards the bottom of the page you will see the `Artifacts` Section, These are the things the build produced again for some reason YOU MUST BE LOGGED INTO GITHUB or you will not be able to download them (its a github thing not us).

## Using the files

Each artifact is an archive that contains one or more builds. Extract the archive and you will have an installer or another archive with the built files.  (Depends on the os). All Builds currently are built in debug mode and The ubuntu 22-04 build need libei 1.2.1+.  While we do not try to land bad code and have some unit tests to help prevent this. It can still happen so running the most recent master you may find some bugs that are not in the current released stable version (when this happens). Builds from Pull Request Branches Should be considered More likely to be Unstable than builds from the master branch.

### I've found a bug what todo
Before you report any bugs, Always try to trigger the bug using the most recent build from the master branch. This helps us greatly many times users are running a few commits behind and we have already fixed the issue.  If the Issue is still happening then you have a bug, Start by checking the issue tracker to see if its been reported We might already know about or even have Pull Request to fix the bug. If we have a PR that should fix your bug try that and report and Let us know on the Pull Request if it worked for you or not.  

#### Bug reports on `master` branches builds
These issues should be reported directly to [Issues](https://github.com/input-leap/input-leap/issues). Be sure to say where you got the build and do note your set up and that you have tested this in the current code in master. 

#### Bugs reports for Pull Request Builds.
If your using a Pull Request and you find a bug. You first have to find if the issue is in the master code or exclusive to the Pull Request, After checking for the bug in master You should also make sure you have tested in the most recent build for the Pull Request. Only than can you consider where to report the issue. If the issues is in the master build report it to Issues. If you find the issue only happens with that Pull Request's build. You should report this on the Pull Request NOT in Issues.

**Reply by the-spyke** [reply]
I see there's `bin/input-leap-flatpak`. Is there a flatpak reference to install somewhere?

**Reply by shymega** [reply]
@the-spyke No, not yet.

**Reply by axelson** [reply]
Awesome!! 🙌

Perhaps unsurprisingly the input-leap-ubuntu-22-04 binary didn't work on my Linux Mint 21.3 machine (it printed an endless loop of `("-f", "--no-tray", "--debug", "INFO", "--name", "linux-mint-desktop", "-c", "/tmp/InputLeap.RZhbLw", "--address", ":24800")` without any other logs). _But_ after compiling from source (https://github.com/input-leap/input-leap/wiki/Building-on-Linux) I have input-leap successfully running on both my Linux Mint desktop and my m1 macBook air (I'm using the `macOS-Apple_Silicon-debug ` binary) 🎉🎉🎉

I'd like to give a BIG thank you to all of the maintainers! :heart:

**Reply by sithlord48** [reply]
The ubuntu 22-04 binaries need libei  try the ubuntu 20 ones (added a note to above)

**Reply by ylluminate** [reply]
Please consider a Homebrew `--HEAD` release to make this simple for macOS users.

**Reply by shymega** [reply]
> Please consider a Homebrew `--HEAD` release to make this simple for macOS users.

That is outside of our scope, in my opinion. I would rather have the Brew community handle packaging. We can provide binaries, but if we start supporting every package manager, it _will_ become quickly overwhelming.

**Reply by ylluminate** [reply]
@shymega I understand what you're saying, but realistically it will not happen. The Homebrew folk have a horrible history of scaling back the recipes and basically have instructed folks to make "taps". Their advice to Leap would be to make a tap to handle this... So unfortunately, if you do absolutely feel this is out of scope, this will not happen unless a Leap community member picks up the torch and has the bandwidth to add it to the build process. You are right that not every package manager should be supported, but macOS and Windows definitely should be supported since they cover a huge set of the community and certain types of users. The Linux community is far more "forgiving" and "accepting" in these regards.

**Reply by the-spyke** [reply]
@shymega But it is a good idea to support 1 major distribution source per platform. For Linux it is Flatpak on Flathub, for macOS it is Homebrew.

**Reply by sithlord48** [reply]
Sounds like there are two users who are interested in maintaining the homebrew port. You can check out our CI script it has everything to make script for homebrew. Our Binaries are released on github that's the only place WE build binaries.

**Reply by Johnhersh** [reply]
Can't get the Mac ones working. I get this error in the logs repeating infinitely:

```
/Users/runner/work/input-leap/input-leap/input-leap/src/lib/inputleap/App.cpp,175
[2024-03-08T20:37:23] FATAL: An error occurred: assistive devices does not trust this process, allow it in system settings.
	/Users/runner/work/input-leap/input-leap/input-leap/src/lib/inputleap/App.cpp,115

[2024-03-08T20:37:23] ERROR: process exited with error code: 1
[2024-03-08T20:37:23] INFO: detected process not running, auto restarting
```

**Reply by the-spyke** [reply]
@Johnhersh have you tried?
```sh
$ tccutil reset Accessibility input-leap
```

**Reply by sithlord48** [reply]
On my mac when i add a new version of inputleap i have to go into the accessibility settings and remove the old one before the new one will work correctly.

**Reply by Johnhersh** [reply]
> @Johnhersh have you tried?
> 
> ```shell
> $ tccutil reset Accessibility input-leap
> ```

Tried it, it doesn't change anything. It looks to do the same as removing it from the Accessibility security list.

**Reply by Johnhersh** [reply]
> On my mac when i add a new version of inputleap i have to go into the accessibility settings and remove the old one before the new one will work correctly.

This is actually on a new Mac. But I tried removing it from Privacy & Security and adding it either automatically or manually even. Still the same result.

**Reply by sithlord48** [reply]
Make sure you have allowed the application to run as well https://support.apple.com/guide/mac-help/open-a-mac-app-from-an-unidentified-developer-mh40616/mac Also make sure you get the correct build for you system

**Reply by Johnhersh** [reply]
Yup, done all that. This is the apple silicon build. Maybe that one has some issues?

**Reply by the-spyke** [reply]
@Johnhersh I'm running the client on M1 and everything is fine after a couple of removals from Privacy and tccutil.

**Reply by Johnhersh** [reply]
No idea, I tried a bunch of times and it fails with the same error as above. What stands out to me though is this path doesn't seem to be correct:
```
/Users/runner/work/input-leap/input-leap/input-leap/src/lib/inputleap/App.cpp
```

There's nothing at that path.

**Reply by shymega** [reply]
@Johnhersh My recommendation is that you take the support request to the 'Discussions' tab - *not* issues - the path you see is expected. It's built on the VM GitHub provides.

**Reply by Jeffrey04** [reply]
> The ubuntu 22-04 build need libei

is `libei-dev` enough (also what version should it be at)? i am getting a lot of ```undefined symbol xdp_input_capture_session_release``` error in ubuntu as client. OTOH, i can't get the ubuntu workstation as host, as the macOS client keeps complaining that it can't connect

anyway managed to get macOS client running, and it can connect to barrier running on ubuntu host (still on x), so partial win for now

**Reply by sithlord48** [reply]
you should install to the `libei` package, The builder uses `v1.2.1` so you will need that version or higher. This is not the place for bug reports so if your issues persists open a proper Issue.

**Reply by Jeffrey04** [reply]
the library needed is actually contained in `libei1` package, and `libei-dev` pulls the package as a dependency on install. Anyway, looks like I will have to wait until 24.04 is out, as that version ships with `1.2.1`. Hopefully everything works fine then, thanks for the help.

Also in order to use input-leap as server on ubuntu, i need to figure out a way to accept incoming connection, so the clients can connect properly, i remember seeing an issue for this

EDIT: something is still broken after upgrading to 24.04, will see if i can find time to submit a bug report over the weekend

**Reply by Jeffrey04** [reply]
just picked one of the latest builds from actions, and compiled the latest main branch from libportal, and input-leap works in my ubuntu 24.04 box as server and client, but still can't get it working in wayland, as i am seeing a lot of these in server log

```
[2024-06-27T17:29:01] ERROR: Failed to initialize InputCapture session, quitting: CreateSession failed
        /__w/input-leap/input-leap/input-leap/src/lib/platform/PortalInputCapture.cpp,136
[2024-06-27T17:29:01] NOTE: disconnecting client "gian.local"
        /__w/input-leap/input-leap/input-leap/src/lib/server/Server.cpp,2085

[2024-06-27T17:29:04] NOTE: stopped server
        /__w/input-leap/input-leap/input-leap/src/lib/inputleap/ServerApp.cpp,768
[2024-06-27T17:29:04] INFO: process exited normally
[2024-06-27T17:29:04] INFO: detected process not running, auto restarting
```

is this a bug, or i am still missing dependencies?

EDIT: no idea what happened, I was having other problems with wayland, and solved it with https://askubuntu.com/a/1516602/70531 and then it also fixed input-leap

**Reply by berglh** [reply]
> just picked one of the latest builds from actions, and compiled the latest main branch from libportal, and input-leap works in my ubuntu 24.04 box as server and client

Hi @Jeffrey04, can you explain the libportal requirement? Going to upgrade my machines and wanting to understand if I can finally switch over to Wayland on 24.04.

**Reply by Jeffrey04** [reply]
@berglh sorry for the late reply, to enable input-leap to work i compiled libportal https://github.com/flatpak/libportal and update the desktop file to point `LD_LIBRARY_PATH` to the compiled library, and things would just work

I ended up switching back to x, as the wayland port would prompt for permission whenever i leap to another machine after certain period (or sometimes it would just freeze)

----

I didn't want to pull all the requirements to my machine, so I compiled the library in a docker container. i then apt-get the required dependencies whenever it vomits an error, but I didn't jot them down. Then i stowed the built library files to `.local`, and patched input-leap's desktop file by replacing the `Exec` configuration

```
Exec=bash -c "LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$HOME/.local/lib input-leap"
```

until https://github.com/flatpak/libportal/issues/152#issuecomment-2309287377 is closed and ubuntu publishes the new version, i suppose we would need to compile libportal in order to use input-leap

YMMV

**Comment by gerroon** [comment]
Where do I download the windows and the linux builds?  I looked at the actions tab and I do not know what I am looking at? Are theyre direct download links? I have been using synergy/barrier for years and I am a good tester.

**Reply by sithlord48** [reply]
This is explained above. You are looking at list of each action that the project has run.
An action is a grouping of Jobs (i.e CI build) . Each action will produce some Artifacts (build files)

**Comment by ftoledo** [comment]
I just download latest build sems to not gzipped:

$ unzip input-leap-ubuntu-22-04.zip 
Archive:  input-leap-ubuntu-22-04.zip
  inflating: input-leap-ubuntu-22-04.tar.gz  
$ tar xvzf input-leap-ubuntu-22-04.tar.gz 

gzip: stdin: not in gzip format
tar: Child returned status 1
tar: Error is not recoverable: exiting now

$ tar xvf input-leap-ubuntu-22-04.tar 
input-leap-ubuntu-22-04/
input-leap-ubuntu-22-04/share/
input-leap-ubuntu-22-04/share/applications/
input-leap-ubuntu-22-04/share/applications/io.github.input_leap.InputLeap.desktop
input-leap-ubuntu-22-04/share/icons/
input-leap-ubuntu-22-04/share/icons/hicolor/
input-leap-ubuntu-22-04/share/icons/hicolor/scalable/
input-leap-ubuntu-22-04/share/icons/hicolor/scalable/apps/
input-leap-ubuntu-22-04/share/icons/hicolor/scalable/apps/io.github.input_leap.InputLeap.svg
input-leap-ubuntu-22-04/share/metainfo/
input-leap-ubuntu-22-04/share/metainfo/io.github.input_leap.InputLeap.appdata.xml
input-leap-ubuntu-22-04/share/man/
input-leap-ubuntu-22-04/share/man/man1/
input-leap-ubuntu-22-04/share/man/man1/input-leaps.1
input-leap-ubuntu-22-04/share/man/man1/input-leapc.1
input-leap-ubuntu-22-04/bin/
input-leap-ubuntu-22-04/bin/input-leaps
input-leap-ubuntu-22-04/bin/input-leapc
input-leap-ubuntu-22-04/bin/input-leap

**Reply by shymega** [reply]
This isn't the place for bug reports. Kindly open a bug in the _proper place_ for issues of this kind.

**Comment by Lex-DRL** [comment]
Maybe, add at least *SOME* note on the current way of getting builds into readme?

From a newcomer's perspective, the repo looks as abandoned as barrier's, but it also seems to have no builds at all. There are releases from 2021, but they have only source code, so a regular windows user looks at this and thinks: "yeah, Linux-only app". A more dedicated windows user would go to wiki, see build instructions page and think: "Ohhh... THAT kind of open-source app" (we don't expect a regular user would read wiki thoroughly, without any chapter called "install", just to see a single line that suggests checking out actions tab not even linking to it, do we?).
Only the most dedicated ones would even get to this discussion and see @sithlord48's post to realize that there actually are builds available, but they're hidden behind actions tab (completely unfamiliar to an overwhelming majority of users who aren't CI-familiar devs themselves and most likely feel really uncomfortable when they see jargon like "artifacts" and "debug").

So... maybe, at least link [the post](https://github.com/input-leap/input-leap/discussions/1793#discussioncomment-8577620) in readme?

**Reply by the-spyke** [reply]
The Curse of Synergy is still going strong :shrug: Please submit a pull request with the readme updated :)

**Reply by Lex-DRL** [reply]
#1911

**Comment by rdmartins** [comment]
Hello guys!

I want to use this with a Windows and a Fedora machine.

On Fedora I'm using without any issues, RPM from the official repo.

On Windows I was using Barrier and want to migrate to InputLeap.
I removed Barrier using the Windows uninstall wizard and downloaded the binaries from the job action artifact.
Now I'm having IPC connection error, connection refused.
I added the three EXEs to the firewall, GUI, server and client, and the error persists.

Should I have something running as a service?

Do you what am I missing here?

Thanks a lot!

**Reply by sithlord48** [reply]
this is not the place for bug reports

**Reply by rdmartins** [reply]
I was not reporting a bug, I was asking a question on a general discussion, because maybe something was going above my head.
I'll build this myself on a Windows VM and figure it out.

I'm sorry that I did not realized this is a bug, but you don't need to be so rude to people
Toxic behavior will only drive people away from the FOSS community.

**Reply by shymega** [reply]
I didn't see this as a bug report either, more of a question. Whilst it may have belonged in a new Q&A thread on Discussions, I can see why @rdmartins posted it here.

The bug itself is known, and the fix is to install Barrier first, then Input Leap. I do apologise for the experience you had here today. Just to note, Input Leap itself from Git is not stable and may break.

**Reply by rdmartins** [reply]
@shymega I built it and created the installer (could not create the service by hand). The input-leapd.exe is there and everything is working as expected:

![daemon](https://github.com/user-attachments/assets/ab531349-9e89-4d23-a53b-b77ee67b6d3b)

The problems I was having typing special characters through input-leap are gone!
Let's see if the strange cursor behavior will be gone as well.

**Reply by shymega** [reply]
Hope it works out for you. Any issues, feel free to open a new Discussion Q&A (if it's a question), or a bug report if you find any bugs! Thanks for the screenshot, too.

**Comment by KoljaFrahm** [comment]
Is there a way to use the Binary builds with wayland? For me under wayland, it runs on Xwayland

**Reply by shymega** [reply]
It depends on the DE. But this isn't the right place to ask this question, your question is different.

Running under XWayland presents a fluke.

You may want to ask on IRC.

**Reply by KoljaFrahm** [reply]
ok will do that
